src/client.rs

```diff
@@ -625,6 +659,38 @@ pub fn get_block_cid_entry(
 	}
 }
 
+fn layout_from_index(index: Vec<(u32, u32)>, size: u32) -> Vec<(u32, u32)> {
+	let mut layout: Vec<(u32, u32)> = Vec::with_capacity(index.len() + 1);
+	if index.len() == 0 {
+		return vec![(0, size)];
+	}
+	for (i, _) in index.iter().enumerate() {
+		if i == 0 {
+			layout.push((0, index[i].1 - 1))
+		} else if i == index.len() - 1 {
+			layout.push((index[i - 1].0, index[i].1 - index[i - 1].1));
+			layout.push((index[i].0, size - index[i].1 + 1));
+		} else {
+			layout.push((index[i - 1].0, index[i].1 - index[i - 1].1));
+		}
+	}
+	layout
+}
```

Comment:
- Move `if` evaluations conditions outside the loop.
-  `enumerate()` is not needed: 
- As we are not consuming `index`, we could just pass a reference to the slice.
```suggestion
fn layout_from_index(index: &[(u32, u32)], size: u32) -> Vec<(u32, u32)> {
	let mut layout: Vec<(u32, u32)> = Vec::with_capacity(index.len() + 1);
	if index.is_empty() {
		return vec![(0, size)];
	}

	layout.push((0, index[0].1 - 1));
	for i in 1..index.len() - 1 {
		layout.push((index[i - 1].0, index[i].1 - index[i - 1].1));
	}

	if index.len() > 1 {
		let last_idx = index.len() - 1;
		layout.push((
			index[last_idx - 1].0,
			index[last_idx].1 - index[last_idx - 1].1,
		));
		layout.push((index[last_idx].0, size - index[last_idx].1 + 1));
	}

	layout
}
```